File: shared/src/commonMain/swift/KvsExtensions.swift
//
// Created by Santiago Mattiauda on 28/9/25.
//

import Foundation
import KvsStorage

public extension Kvs {
    public func getStringAsResult(key : String, defValue: String) async -> Result<String, Error> {
        do {
            return .success(try await getString(key: key, defValue: defValue))
        } catch {
            return .failure(error)
        }
    }

    public func getIntAsResult(key : String, defValue: Int32) async -> Result<Int32, Error> {
        do {
            return .success(try await getInt(key: key, defValue: defValue).int32Value)
        } catch {
            return .failure(error)
        }
    }

    public func getLongAsResult(key : String, defValue: Int64) async -> Result<Int64, Error> {
        do {
            return .success(try await getLong(key: key, defValue: defValue).int64Value)
        } catch {
            return .failure(error)
        }
    }

    public func getFloatAsResult(key : String, defValue: Float) async -> Result<Float, Error> {
        do {
            return .success(try await getFloat(key: key, defValue: defValue).floatValue)
        } catch {
            return .failure(error)
        }
    }

    public func getBooleanAsResult(key : String, defValue: Bool) async -> Result<Bool, Error> {
        do {
            return .success(try await getBoolean(key: key, defValue: defValue).boolValue)
        } catch {
            return .failure(error)
        }
    }
}

public extension KvsKvsEditor {

    public func apply() async -> Result<Bool, Error>{
        do {
            try await commit()
            return .success(true)
        } catch {
            return .failure(error)
        }
    }
}

File: kvs-core/src/commonMain/kotlin/com/santimattius/kvs/KvsExtensions.kt
package com.santimattius.kvs

import com.santimattius.kvs.internal.extensions.runNonCancellableCatching
import kotlinx.coroutines.CancellationException
import kotlin.experimental.ExperimentalObjCRefinement
import kotlin.native.HiddenFromObjC

@OptIn(ExperimentalObjCRefinement::class)
@HiddenFromObjC
@Throws(CancellationException::class)
suspend fun Kvs.getStringAsResult(key: String, defValue: String): Result<String> =
    runNonCancellableCatching { getString(key, defValue) }

@OptIn(ExperimentalObjCRefinement::class)
@HiddenFromObjC
@Throws(CancellationException::class)
suspend fun Kvs.getIntAsResult(key: String, defValue: Int): Result<Int> =
    runNonCancellableCatching { getInt(key, defValue) }

@OptIn(ExperimentalObjCRefinement::class)
@HiddenFromObjC
@Throws(CancellationException::class)
suspend fun Kvs.getLongAsResult(key: String, defValue: Long): Result<Long> =
    runNonCancellableCatching { getLong(key, defValue) }

suspend fun Kvs.getFloatAsResult(key: String, defValue: Float): Result<Float> =
    runNonCancellableCatching { getFloat(key, defValue) }

@OptIn(ExperimentalObjCRefinement::class)
@HiddenFromObjC
@Throws(CancellationException::class)
suspend fun Kvs.getBooleanAsResult(key: String, defValue: Boolean): Result<Boolean> =
    runNonCancellableCatching { getBoolean(key, defValue) }

@OptIn(ExperimentalObjCRefinement::class)
@HiddenFromObjC
@Throws(CancellationException::class)
suspend fun Kvs.getAllAsResult(): Result<Map<String, Any>> =
    runNonCancellableCatching { getAll() }

@OptIn(ExperimentalObjCRefinement::class)
@HiddenFromObjC
@Throws(CancellationException::class)
suspend fun Kvs.KvsEditor.apply(): Result<Boolean> =
    runNonCancellableCatching { commit(); true }
